### PR TITLE
Add grid-search optimizer

### LINKED ISSUE #85: Add gridsearch optimizer

### PR DESCRIPTION
**Pull request by moritzmeister**
closes #85

### CODE REVIEW

File: maggy/optimizer/gridsearch.py [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+            *[item["values"] for item in searchspace.items()]
+        ):
+            return_list.append(searchspace.list_to_dict(hparams))
+        return return_list
+
+    @classmethod
+    def _validate_searchspace(cls, searchspace):
+        if (
+            Searchspace.DOUBLE in searchspace.names().values()
+            and Searchspace.INTEGER in searchspace.names().values()
```
Comment (amacati) [inline maggy/optimizer/gridsearch.py]: Is this supposed to be a logical "and"? Seems like is's supposed to be an "or".

File: maggy/optimizer/gridsearch.py [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+    def _grid_params(cls, searchspace):
+        return_list = []
+        for hparams in itertools.product(
+            *[item["values"] for item in searchspace.items()]
+        ):
+            return_list.append(searchspace.list_to_dict(hparams))
+        return return_list
+
+    @classmethod
+    def _validate_searchspace(cls, searchspace):
```
Comment (amacati) [inline maggy/optimizer/gridsearch.py]: cls unused, could be staticmethod instead.

File: maggy/optimizer/gridsearch.py [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+
+            return next_trial
+        else:
+            return None
+
+    def finalize_experiment(self, trials):
+        return
+
+    @classmethod
+    def _grid_params(cls, searchspace):
```
Comment (amacati) [inline maggy/optimizer/gridsearch.py]: cls unused, could be staticmethod instead.

**Review by amacati — COMMENTED** [review]
Recheck the logical "and" + class methods in gridsearch.py, otherwise looks fine. 

Nit: I'd argue we should enforce lowerstyle naming conventions more strictly (abstract_optimizer.py instead of abstractoptimizer.py, increases readability and consistency).

File: maggy/optimizer/gridsearch.py [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+        """
+        cls._validate_searchspace(searchspace)
+        return len(cls._grid_params(searchspace))
+
+    def get_suggestion(self, trial=None):
+        # sampling routine for randomsearch + pruner
+        if self.pruner:
+            raise NotImplementedError(
+                "Grid search in combination with trial pruning"
+                "is currently not supported."
```
Comment (RiccardoGrigoletto) [inline maggy/optimizer/gridsearch.py:50]: add whitespace before 'is'.

File: maggy/optimizer/gridsearch.py [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+
+    @classmethod
+    def _validate_searchspace(cls, searchspace):
+        if (
+            Searchspace.DOUBLE in searchspace.names().values()
+            and Searchspace.INTEGER in searchspace.names().values()
+        ):
+            raise NotImplementedError(
+                "Searchspace can only contain `discrete` or `categorical`"
+                "hyperparameters for grid search."
```
Comment (RiccardoGrigoletto) [inline maggy/optimizer/gridsearch.py:89]: add whitespace before 'hyperparamenters'.

File: maggy/optimizer/gridsearch.py (hunk 1 above)
Comment (moritzmeister) [reply maggy/optimizer/gridsearch.py]: thanks

**Review by amacati — APPROVED** [review]
Seems fine.

**Review by RiccardoGrigoletto — APPROVED** [review]

**Comment by moritzmeister** [comment]
rebased on martin's changes

File: maggy/core/experiment_driver/OptimizationDriver.py [hunk 6]
```diff
@@ -18,10 +18,10 @@
 
 from maggy import util
 from maggy.searchspace import Searchspace
-from maggy.optimizer import AbstractOptimizer, RandomSearch, Asha, SingleRun
+from maggy.optimizer import AbstractOptimizer, RandomSearch, Asha, SingleRun, GridSearch
 from maggy.earlystop import AbstractEarlyStop, MedianStoppingRule, NoStoppingRule
 from maggy.optimizer import bayes
-from maggy.core.experiment_driver.Driver import Driver
+from maggy.core.experiment_driver import Driver
```
Comment (amacati) [inline maggy/core/experiment_driver/OptimizationDriver.py]:
```suggestion
from maggy.core.experiment_driver.Driver import Driver
```

**Review by amacati — COMMENTED** [review]
I think you broke the import of the Driver.

**Pull request merged**